File: Asignment2/parser.py
EOF = -1
EQ_OP = 10
NEQ_OP = 11
LS_OP = 20
GR_OP = 21
LE_OP = 22
GE_OP = 23
UNKNOWN = -1

class Parser:
    def parse(self, input:str):
        self.__index = 0
        self.__string = input
        self.__stack = []
        try:
            self.__expr()
        except Exception as e:
            print(f">> {e}")

    def __lookup(self) -> str:
        if self.__index >= len(self.__string): return EOF
        while(self.__string[self.__index] == " "):
            self.__index+=1
        return self.__string[self.__index]

    def __is_digit(self) -> bool: 
        if self.__index >= len(self.__string): return False
        return "0" <= self.__string[self.__index] <= "9"

    def __error(self):
        raise Exception("syntax error!!")

    def __lexeme(self, c:str):
        if self.__index >= len(self.__string): self.__error()
        if self.__string[self.__index] == c:
            self.__index += 1
        else:
            self.__error()

    def __number(self):
        # print("enter __number")
        num = ""
        if not self.__is_digit(): self.__error()
        else:
            while self.__is_digit():
                num += self.__string[self.__index]
                self.__index += 1
            self.__stack.append(int(num))

    def __factor(self):
        # print("enter __factor")
        if self.__lookup() == "(":
            self.__lexeme("(")
            self.__aexpr()
            self.__lexeme(")")
        else:
            self.__number()

    def __term(self):
        # print("enter __term")
        self.__factor()
        while self.__lookup() == "+" or self.__lookup() == "-":
            left = self.__stack.pop()
            if self.__lookup() == "+":
                self.__lexeme(self.__lookup())
                self.__factor()
                right = self.__stack.pop()
                self.__stack.append(left+right)
            elif self.__lookup() == "-":
                self.__lexeme(self.__lookup())
                self.__factor()
                right = self.__stack.pop()
                self.__stack.append(left-right)

    def __aexpr(self):
        # print("enter __aexpr")
        self.__term()
        while self.__lookup() == "*" or self.__lookup() == "/":
            left = self.__stack.pop()
            if self.__lookup() == "*":
                self.__lexeme(self.__lookup())
                self.__term()
                right = self.__stack.pop()
                self.__stack.append(left*right)
            elif self.__lookup() == "/":
                self.__lexeme(self.__lookup())
                self.__term()
                right = self.__stack.pop()
                if(right == 0):
                    raise Exception("divie by zero exception")
                self.__stack.append(left//right)

    def __relop(self) -> int:
        # print("enter __relop")
        if self.__lookup() == "=":
            self.__lexeme("=")
            self.__lexeme("=")
            return EQ_OP
        elif self.__lookup() == "!":
            self.__lexeme("!")
            self.__lexeme("=")
            return NEQ_OP
        elif self.__lookup() == ">":
            self.__lexeme(">")
            if self.__string[self.__index] == "=":
                self.__lexeme("=")
                return GE_OP
            else: return GR_OP
        elif self.__lookup() == "<":
            self.__lexeme("<")
            if self.__string[self.__index] == "=":
                self.__lexeme("=")
                return LE_OP
            else: return LS_OP
        else: self.__error()
        return UNKNOWN

    def __bexpr(self):
        # print("enter __bexpr")
        left = self.__stack.pop()
        rop = self.__relop()
        self.__aexpr()
        right = self.__stack.pop()
        if rop == EQ_OP:
            print(f">> {left == right}")
        elif rop == NEQ_OP:
            print(f">> {left != right}")
        elif rop == LS_OP:
            print(f">> {left < right}")
        elif rop == GR_OP:
            print(f">> {left > right}")
        elif rop == LE_OP:
            print(f">> {left <= right}")
        elif rop == GE_OP:
            print(f">> {left >= right}")

    def __expr(self):
        # print("enter __expr")
        if self.__lookup() == "(" or self.__is_digit():
            self.__aexpr()
        else: self.__error()

        if self.__lookup() != EOF:
            self.__bexpr()
        else:
            print(f">> {self.__stack.pop()}")

parser = Parser()
while(True):
    expr = input()
    if(len(expr) == 0): break
    parser.parse(expr)

File: Asignment2/parser.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include <stack>
#define EQ_OP 10
#define NEQ_OP 11
#define LS_OP 20
#define GR_OP 21
#define LE_OP 22
#define GE_OP 23
#define UNKNOWN -1
using namespace std;

class Parser {
public:
    void parse(string input) {
        index = 0;
        str = input;
        while(!s.empty()) s.pop();
        try{
            expr();
        }
        catch(exception& e){
            cout << e.what();
        }
    }

private:
    stack<int> s;
    string str;
    int index;

    char lookup() {
        if(index >= str.length()) return EOF;
        while(str[index] == ' ') index++;
        return str[index];
    }

    bool is_digit() {
        return str[index] >= '0' && str[index] <= '9';
    }

    void error() {
        throw runtime_error("syntax error!!\n");
    }

    void lexeme(char c) {
        if (str[index] == c) index++;
        else error();
    }

    void number() {
        string num = "";
        if (!is_digit()) error();
        else{
            while (is_digit()) {
                num+=str[index];
                index++;
            }
            s.push(stoi(num));
        }
    }

    void factor() {
        // cout << "enter factor\n";
        if (lookup() == '(') {
            lexeme('(');
            aexpr();
            lexeme(')');
        } else number();
    }

    void term() {
        // cout << "enter term\n";
        factor();
        while (lookup() == '+' || lookup() == '-') {
            int left = s.top();
            s.pop();
            if(lookup() == '+'){
                lexeme(lookup());
                factor();
                int right = s.top();
                s.pop();
                s.push(left+right);
            }
            else if(lookup() == '-'){
                lexeme(lookup());
                factor();
                int right = s.top();
                s.pop();
                s.push(left-right);
            }
        }
    }

    void aexpr() {
        // cout << "enter aexpr\n";
        term();
        while (lookup() == '*' || lookup() == '/') {
            int left = s.top();
            s.pop();
            if(lookup() == '*'){
                lexeme(lookup());
                term();
                int right = s.top();
                s.pop();
                s.push(left*right);
            }
            else if(lookup() == '/'){
                lexeme(lookup());
                term();
                int right = s.top();
                s.pop();
                if(right == 0) throw runtime_error("divide by zero exception\n");
                s.push(left/right);
            }
        }
    }

    int relop() {
        // cout << "enter relop\n";
        if(lookup() == '='){
            lexeme('=');
            lexeme('=');
            return EQ_OP;
        }
        else if(lookup() == '!'){
            lexeme('!');
            lexeme('=');
            return NEQ_OP;
        }
        else if(lookup() == '>'){
            lexeme('>');
            if(str[index] == '='){
                lexeme('=');
                return GE_OP;
            }
            else return GR_OP;
        }
        else if(lookup() == '<'){
            lexeme('<');
            if(str[index] == '='){
                lexeme('=');
                return LE_OP;
            }
            else return LS_OP;
        }
        else error();
        return UNKNOWN;
    }

    void bexpr() {
        // cout << "enter bexpr\n";
        int left = s.top();
        s.pop();
        int rop = relop();
        aexpr();
        int right = s.top();
        s.pop();
        switch(rop){
            case EQ_OP:
                cout << ">> " << (left == right ? "true" : "false") << '\n';
                break;
            case NEQ_OP:
                cout << ">> " << (left != right ? "true" : "false") << '\n';
                break;
            case LS_OP:
                cout << ">> " << (left < right ? "true" : "false") << '\n';
                break;
            case GR_OP:
                cout << ">> " << (left > right ? "true" : "false") << '\n';
                break;
            case LE_OP:
                cout << ">> " << (left <= right ? "true" : "false") << '\n';
                break;
            case GE_OP:
                cout << ">> " << (left >= right ? "true" : "false") << '\n';
                break;
            default:
                error();
        }
    }

    void expr() {
        // cout << "enter expr\n";
        if (lookup() == '(' || is_digit()) {
            aexpr();
        }
        else error();
        if(lookup() != EOF){
            bexpr();
        }
        else cout << ">> " << s.top() << '\n';
    }
};

int main() {
    Parser parser;
    string input;
    while(true){
        getline(cin, input);
        if(input.length() == 0) break;
            parser.parse(input);
    }
}
